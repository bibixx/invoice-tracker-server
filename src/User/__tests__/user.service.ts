import sinon from "sinon";

import UserModel from "../user.model";
import UserService from "../user.service";

const User = <any>UserModel;

const password = "Abcdefghijk123"; // cSpell:disable-line

describe("UserService.createUser", () => {
  beforeEach(() => {
    sinon.stub(User, "findOne");
    sinon.stub(User, "create");
  });

  afterEach(() => {
    User.findOne.restore();
    User.create.restore();
  });

  it("should create a new user", async () => {
    User.findOne.resolves(null);
    User.create.resolves({});

    const US = UserService(User);

    const res = await US.createUser("a@a.pl", password, password);
    expect(res).not.toThrow;
    expect(sinon.assert.calledOnce(User.create));
  });

  it("should throw error if passwords don't match", async () => {
    User.findOne.resolves(null);
    User.create.resolves({});

    const US = UserService(User);
    await expect(US.createUser("a@a.pl", password, "b")).rejects.toThrowErrorMatchingSnapshot();
    expect(sinon.assert.notCalled(User.create));
  });

  it("should throw error if email is invalid", async () => {
    User.findOne.resolves(null);
    User.create.resolves({});

    const US = UserService(User);
    await expect(US.createUser("a@a", password, password)).rejects.toThrowErrorMatchingSnapshot();
    expect(sinon.assert.notCalled(User.create));
  });

  it("should throw error if user was previously signed up", async () => {
    User.findOne.resolves(true);
    User.create.resolves({});

    const US = UserService(User);

    await expect(US.createUser("a@a.pl", password, password))
      .rejects.toThrowErrorMatchingSnapshot();
    expect(sinon.assert.notCalled(User.create));
  });

  it("should throw error if password doesn't match requirements", async () => {
    User.findOne.resolves(null);
    User.create.resolves({});

    const US = UserService(User);

    /* cSpell:disable */
    await expect(
      US.createUser("a@a.pl", "ą", "ą"),
    ).rejects.toThrowErrorMatchingSnapshot();

    await expect(
      US.createUser("a@a.pl", "ąbcdefgh", "ąbcdefgh"),
    ).rejects.toThrowErrorMatchingSnapshot();

    await expect(
      US.createUser("a@a.pl", "Ąbcdefgh", "Ąbcdefgh"),
    ).rejects.toThrowErrorMatchingSnapshot();

    await expect(
      US.createUser("a@a.pl", "Ąbcdefgh1", "Ąbcdefgh1"),
    ).resolves.not.toThrow();

    /* cSpell:enable */

    expect(sinon.assert.calledOnce(User.create));
  });
});

describe("UserService.login", () => {
  beforeEach(() => {
    sinon.stub(User, "findOne");
  });

  afterEach(() => {
    User.findOne.restore();
  });

  it("should throw error if user doesn't exist", async () => {
    User.findOne.resolves(null);

    const US = UserService(User);

    await expect(US.login("a@a.pl", password)).rejects.toThrowErrorMatchingSnapshot();
  });

  it("should return error if password is invalid", async () => {
    User.findOne.resolves({
      comparePasswords: () => Promise.resolve(false),
    });

    const US = UserService(User);

    await expect(US.login("a@a.pl", password)).rejects.toThrowErrorMatchingSnapshot();
  });

  it("should successfully login", async () => {
    User.findOne.resolves({
      email: true,
      comparePasswords: () => Promise.resolve(true),
    });

    const US = UserService(User);

    await expect(US.login("a@a.pl", password)).resolves.not.toThrow();
  });

  it("should throw error if password doesn't match requirements", async () => {
    User.findOne.resolves({
      comparePasswords: () => Promise.resolve(true),
    });

    const US = UserService(User);

    /* cSpell:disable */
    await expect(US.login("a@a.pl", "ą")).rejects.toThrowErrorMatchingSnapshot();
    await expect(US.login("a@a.pl", "ąbcdefgh")).rejects.toThrowErrorMatchingSnapshot();
    await expect(US.login("a@a.pl", "Ąbcdefgh")).rejects.toThrowErrorMatchingSnapshot();
    await expect(US.login("a@a.pl", "Ąbcdefgh1")).resolves.not.toThrow();
    /* cSpell:enable */
  });
});
