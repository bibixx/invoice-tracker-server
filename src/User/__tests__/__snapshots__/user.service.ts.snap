// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserService.createUser should throw error if email is invalid 1`] = `"email must be a valid email"`;

exports[`UserService.createUser should throw error if password doesn't match requirements 1`] = `"password must be at least 8 characters"`;

exports[`UserService.createUser should throw error if password doesn't match requirements 2`] = `"password must contain uppercase letter"`;

exports[`UserService.createUser should throw error if password doesn't match requirements 3`] = `"password must contain a number"`;

exports[`UserService.createUser should throw error if passwords don't match 1`] = `"passwords must match"`;

exports[`UserService.createUser should throw error if user was previously signed up 1`] = `"user with email a@a.pl was already created"`;

exports[`UserService.login should return error if password is invalid 1`] = `"incorrect email or password."`;

exports[`UserService.login should throw error if password doesn't match requirements 1`] = `"password must be at least 8 characters"`;

exports[`UserService.login should throw error if password doesn't match requirements 2`] = `"password must contain uppercase letter"`;

exports[`UserService.login should throw error if password doesn't match requirements 3`] = `"password must contain a number"`;

exports[`UserService.login should throw error if user doesn't exist 1`] = `"incorrect email or password."`;
